src/Aspire.Hosting/ApplicationModel/HostUrl.cs

```diff
@@ -44,25 +41,70 @@ public record HostUrl(string Url) : IValueProvider, IManifestExpressionProvider
             var uri = new UriBuilder(Url);
             if (uri.Host is "localhost" or "127.0.0.1" or "[::1]")
             {
-                var hasEndingSlash = Url.EndsWith('/');
-                uri.Host = KnownHostNames.DefaultContainerTunnelHostName;
-                retval = uri.ToString();
-
-                // Remove trailing slash if we didn't have one before (UriBuilder always adds one)
-                if (!hasEndingSlash && retval.EndsWith('/'))
+                if (context.ExecutionContext?.IsRunMode == true)
                 {
-                    retval = retval[..^1];
+                    var options = context.ExecutionContext.ServiceProvider.GetRequiredService<IOptions<DcpOptions>>();
+
+                    var infoService = context.ExecutionContext.ServiceProvider.GetRequiredService<IDcpDependencyCheckService>();
+                    var dcpInfo = await infoService.GetDcpInfoAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                    uri.Host = options.Value.EnableAspireContainerTunnel? KnownHostNames.DefaultContainerTunnelHostName : dcpInfo?.Containers?.ContainerHostName ?? KnownHostNames.DockerDesktopHostBridge;
+
+                    if (options.Value.EnableAspireContainerTunnel)
+                    {
+                        // We need to consider that both the host and port may need to be remapped
+                        var model = context.ExecutionContext.ServiceProvider.GetRequiredService<DistributedApplicationModel>();
+                        var targetEndpoint = model.Resources.Where(r => !r.IsContainer())
```

Comment:
Why do we need to scan the entire model??

Reply:
All we know in HostUrl is an address and port, but both the address and port may need to change if running in a container. So I'm finding out if it's the URL to an existing endpoint and then mapping that to the appropriate container network endpoint if one exists. Ideally we'd just have a way to get the endpoints from the dashboard directly here and deprecate HostUrl entirely, but that's a larger change.